librz/bin/pdb/gdata.c

```diff
@@ -1,84 +1,94 @@
+		rz_buf_seek(buf, remain, RZ_BUF_CUR);
+		read_len += remain;
+	}
+	return true;
 }
 
-void parse_gdata_stream(void *stream, RZ_STREAM_FILE *stream_file) {
-	unsigned short len = 0;
-	unsigned short leaf_type = 0;
-	char *data = 0;
-	SGDATAStream *data_stream = (SGDATAStream *)stream;
-	SGlobal *global = 0;
-
-	data_stream->globals_list = rz_list_new();
-	while (1) {
-		stream_file_read(stream_file, 2, (char *)&len);
-		if (len == 0) {
+RZ_IPI bool parse_gdata_stream(RzPdb *pdb, MsfStream *stream) {
+	rz_return_val_if_fail(pdb && stream, false);
+	if (!pdb->s_gdata) {
+		pdb->s_gdata = RZ_NEW0(GDataStream);
+	}
+	GDataStream *s = pdb->s_gdata;
+	if (!s) {
+		RZ_LOG_ERROR("Error allocating memory.\n");
+		return false;
+	}
+	RzBuffer *buf = stream->stream_data;
+	s->global_list = rz_list_new();
+	if (!s->global_list) {
+		return false;
+	}
+	ut16 len;
+	while (true) {
+		ut32 read_len = 0;
+		if (!rz_buf_read_le16(buf, &len)) {
+			break;
+		}
+		read_len += sizeof(ut16);
+		if (len == 0 || len == UT16_MAX) {
 			break;
 		}
-		data = (char *)malloc(len);
-		if (!data) {
-			return;
+		ut16 leaf_type;
+		if (!rz_buf_read_le16(buf, &leaf_type)) {
+			return false;
 		}
-		stream_file_read(stream_file, len, data);
-
-		leaf_type = *(unsigned short *)(data);
-		if ((leaf_type == 0x110E) || (leaf_type == 0x1009)) {
-			global = (SGlobal *)malloc(sizeof(SGlobal));
+		read_len += sizeof(ut16);
+		if (leaf_type == 0x110E || leaf_type == 0x1009) {
+			GDataGlobal *global = RZ_NEW0(GDataGlobal);
 			if (!global) {
-				free(data);
-				return;
+				goto skip;
 			}
 			global->leaf_type = leaf_type;
-			parse_global(data + 2, len, global);
-			rz_list_append(data_stream->globals_list, global);
+			if (!parse_gdata_global(global, buf, &read_len)) {
+				RZ_FREE(global);
+				return false;
+			}
+			rz_list_append(s->global_list, global);
+			continue;
 		}
-		free(data);
+	skip:
+		rz_buf_seek(buf, len - sizeof(ut16), RZ_BUF_CUR);
 	}
-
-	// TODO: for more fast access
-	//	for g in self.globals:
-	//        if not hasattr(g, 'symtype'): continue
-	//        if g.symtype == 0:
-	//            if g.name.startswith("_"):
-	//                self.vars[g.name[1:]] = g
-	//            else:
-	//                self.vars[g.name] = g
-	//        elif g.symtype == 2:
-	//            self.funcs[g.name] = g
+	return true;
 }
 
-void free_gdata_stream(void *stream) {
-	SGDATAStream *data_stream = (SGDATAStream *)stream;
-	SGlobal *global = 0;
-	RzListIter *it = 0;
-
-	it = rz_list_iterator(data_stream->globals_list);
-	while (rz_list_iter_next(it)) {
-		global = (SGlobal *)rz_list_iter_get(it);
-		if (global->name.name) {
-			free(global->name.name);
-		}
-		free(global);
+RZ_IPI void free_gdata_stream(GDataStream *stream) {
+	GDataGlobal *global;
+	RzListIter *it;
+	rz_list_foreach (stream->global_list, it, global) {
+		RZ_FREE(global->name);
+		RZ_FREE(global);
 	}
-	rz_list_free(data_stream->globals_list);
+	rz_list_free(stream->global_list);
 }
```

Comment:
To be honest, I wrote it this way on purpose, because there is only this one list in this stream, which obviously makes the logic clearer. If I use `rz_list_newf()`, then I just call `rz_list_free(s_gdata->global_list)` once in `rz_pdb_free()`, but this is not good for understanding, I think.